File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/containers/ScrollContainerView.kt
package com.crescentflare.piratesgame.components.containers

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.os.Handler
import android.util.AttributeSet
import android.view.View
import com.crescentflare.piratesgame.components.utility.ViewletUtil
import com.crescentflare.piratesgame.infrastructure.events.AppEvent
import com.crescentflare.piratesgame.infrastructure.events.AppEventObserver
import com.crescentflare.piratesgame.infrastructure.events.AppEventType
import com.crescentflare.unilayout.containers.UniVerticalScrollContainer
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.piratesgame.infrastructure.inflator.Inflators

import java.lang.ref.WeakReference

/**
 * Container view: a scroll view specifically made for the layout containers (like LinearContainerView)
 */
class ScrollContainerView : UniVerticalScrollContainer, AppEventObserver {

    // --
    // Static: viewlet integration
    // --

    companion object {

        val viewlet: JsonInflatable = object : JsonInflatable {

            override fun create(context: Context): Any {
                return ScrollContainerView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is ScrollContainerView) {
                    // Set fill content mode
                    obj.isFillViewport = mapUtil.optionalBoolean(attributes, "fillContent", false)

                    // Set content
                    val item = Inflators.viewlet.attributesForNestedInflatable(attributes["item"])
                    val recycling = mapUtil.optionalBoolean(attributes, "recycling", false)
                    if (recycling && Inflators.viewlet.canRecycle(obj.contentView, item)) {
                        val contentView = obj.contentView
                        if (contentView != null && item != null) {
                            Inflators.viewlet.inflateOn(contentView, item, null)
                            ViewletUtil.applyLayoutAttributes(mapUtil, contentView, item)
                            ViewletUtil.bindRef(mapUtil, contentView, item, binder)
                        }
                    } else {
                        // First empty content
                        obj.contentView = null

                        // Set content view
                        val result = Inflators.viewlet.inflate(obj.context, item, obj, binder)
                        if (result is View && item != null) {
                            obj.addView(result)
                            ViewletUtil.applyLayoutAttributes(mapUtil, result, item)
                            ViewletUtil.bindRef(mapUtil, result, item, binder)
                        }
                    }

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)

                    // Forward event observer
                    if (parent is AppEventObserver) {
                        obj.eventObserver = parent
                    }

                    // Set pull to refresh (after interaction listener is set)
                    obj.pullToRefreshEvent = AppEvent.fromObject(attributes["pullToRefreshEvent"])
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj is ScrollContainerView
            }

        }

    }


    // --
    // Members
    // --

    private var eventObserverReference : WeakReference<AppEventObserver>? = null
    private var atScrollTop = false
    private var lastScroll = -1


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
            context: Context,
            attrs: AttributeSet? = null,
            defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
            context: Context,
            attrs: AttributeSet?,
            defStyleAttr: Int,
            defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        clipToPadding = false
        Handler().postDelayed({
            onScrollStartChanged(scrollY == 0)
            lastScroll = scrollY
        }, 0)
    }


    // --
    // Configurable values
    // --

    var eventObserver: AppEventObserver?
        get() = eventObserverReference?.get()
        set(newValue) {
            eventObserverReference = if (newValue != null) {
                WeakReference(newValue)
            } else {
                null
            }
        }

    var pullToRefreshEvent: AppEvent? = null
        set(pullToRefreshEvent) {
            if (field != pullToRefreshEvent) {
                if (field != null && atScrollTop) {
                    val event = AppEvent(AppEventType.RegisterPullToRefresh.value + "://remove")
                    eventObserver?.observedEvent(event, field)
                }
                field = pullToRefreshEvent
                if (pullToRefreshEvent != null && atScrollTop) {
                    val event = AppEvent(AppEventType.RegisterPullToRefresh.value + "://add")
                    eventObserver?.observedEvent(event, field)
                }
            }
        }

    var contentView: View?
        get() {
            val childCount = childCount
            return if (childCount > 0) {
                getChildAt(0)
            } else null
        }
        set(contentView) {
            val childCount = childCount
            for (i in childCount - 1 downTo 0) {
                removeViewAt(i)
            }
            if (contentView != null) {
                addView(contentView, 0)
            }
        }


    // --
    // Interaction
    // --

    fun onScrollStartChanged(atScrollStart: Boolean) {
        atScrollTop = atScrollStart
        if (pullToRefreshEvent != null) {
            if (atScrollTop) {
                val event = AppEvent(AppEventType.RegisterPullToRefresh.value + "://add")
                eventObserver?.observedEvent(event, pullToRefreshEvent)
            } else {
                val event = AppEvent(AppEventType.RegisterPullToRefresh.value + "://remove")
                eventObserver?.observedEvent(event, pullToRefreshEvent)
            }
        }
    }

    override fun onScrollChanged(left: Int, top: Int, previousLeft: Int, previousTop: Int) {
        super.onScrollChanged(left, top, previousLeft, previousTop)
        if (top == 0 && lastScroll != 0) {
            onScrollStartChanged(true)
        } else if (top != 0 && lastScroll == 0) {
            onScrollStartChanged(false)
        }
        lastScroll = top
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        if (pullToRefreshEvent != null && atScrollTop) {
            val event = AppEvent(AppEventType.RegisterPullToRefresh.value + "://remove")
            eventObserver?.observedEvent(event, pullToRefreshEvent)
        }
    }

    override fun observedEvent(event: AppEvent, sender: Any?) {
        eventObserver?.observedEvent(event, sender)
    }

}

File: pirates-ios/PiratesGame/Components/Containers/ScrollContainerView.swift
//
//  ScrollContainerView.swift
//  Container view: a scroll view specifically made for the layout containers (like LinearContainerView)
//

import UIKit
import UniLayout
import JsonInflator

class ScrollContainerView: UniVerticalScrollContainer, AppEventObserver {
    
    // --
    // MARK: Members
    // --
    
    internal var refreshControlView: UIRefreshControl?
    
    
    // --
    // MARK: Viewlet integration
    // --
    
    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return ScrollContainerView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let scrollContainer = object as? ScrollContainerView {
                // Set fill content mode
                scrollContainer.fillContent = convUtil.asBool(value: attributes["fillContent"]) ?? false
                
                // Update content view
                let checkItem = Inflators.viewlet.attributesForNestedInflatable(attributes["item"])
                let recycling = convUtil.asBool(value: attributes["recycling"]) ?? false
                if recycling && Inflators.viewlet.canRecycle(object: scrollContainer.contentView, attributes: checkItem) {
                    if let item = checkItem {
                        if let view = scrollContainer.contentView {
                            Inflators.viewlet.inflate(onObject: view, attributes: item)
                            ViewletUtil.applyLayoutAttributes(convUtil: convUtil, view: view, attributes: item)
                            ViewletUtil.bindRef(view: view, attributes: item, binder: binder)
                        }
                    }
                } else {
                    // First empty content
                    scrollContainer.contentView = nil
                    
                    // Set content item
                    if let item = checkItem {
                        if let view = Inflators.viewlet.inflate(attributes: item, parent: scrollContainer, binder: binder) as? UIView {
                            scrollContainer.contentView = view
                            ViewletUtil.applyLayoutAttributes(convUtil: convUtil, view: view, attributes: item)
                            ViewletUtil.bindRef(view: view, attributes: item, binder: binder)
                        }
                    }
                }
                
                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: scrollContainer, attributes: attributes)
                
                // Event handling
                scrollContainer.pullToRefreshEvent = AppEvent(value: attributes["pullToRefreshEvent"])

                // Forward event observer
                if let eventObserver = parent as? AppEventObserver {
                    scrollContainer.eventObserver = eventObserver
                }
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return object is ScrollContainerView
        }
        
    }
    
    
    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    private func setup() {
        if #available(iOS 11.0, *) {
            contentInsetAdjustmentBehavior = .never
        }
    }
    
    
    // --
    // MARK: Configurable values
    // --

    weak var eventObserver: AppEventObserver?

    var pullToRefreshEvent: AppEvent? {
        didSet {
            if pullToRefreshEvent != nil && refreshControlView == nil {
                refreshControlView = UIRefreshControl()
                refreshControlView?.tintColor = AppColors.primary
                refreshControlView?.addTarget(self, action: #selector(pulledToRefresh(_:)), for: .valueChanged)
                backgroundView = refreshControlView
                alwaysBounceVertical = true
            } else if pullToRefreshEvent == nil && refreshControlView != nil {
                backgroundView = nil
                refreshControlView = nil
                alwaysBounceVertical = false
            }
        }
    }
    
    func stopRefreshing() {
        refreshControlView?.endRefreshing()
    }
    
    
    // --
    // MARK: Interaction
    // --
    
    func observedEvent(_ event: AppEvent, sender: Any?) {
        eventObserver?.observedEvent(event, sender: sender)
    }

    @objc func pulledToRefresh(_ refreshControl: UIRefreshControl) {
        if let event = pullToRefreshEvent {
            eventObserver?.observedEvent(event, sender: self)
        }
    }

}
